### PR TITLE
 [MP Rewrite Part 2] Introduce the Trajectory abstraction

### PR DESCRIPTION
**Pull request by kylestach**
A Trajectory is a piecewise-cubic function interpolating between a
vector of time-indexed RobotInstant waypoints.

This is not expected to pass CI.

### CODE REVIEW

File: common/Utils.hpp [hunk 1]
```diff
@@ -61,10 +61,7 @@ inline void debugThrowIf(const std::string& string, bool condition) {
  */
 template <typename T>
 static inline T fixAngleRadians(T a) {
-    a = remainder(a, 2 * M_PI);
-    while (a < -M_PI) a += 2.0 * M_PI;
-    while (a > M_PI) a -= 2.0 * M_PI;
-    return a;
+    return std::atan2(std::sin(a), std::cos(a));
```
Comment (oswinso) [inline common/Utils.hpp]: Why this change?

File: soccer/motion/MotionControl.cpp [hunk 2]
```diff
@@ -102,7 +98,7 @@ void MotionControl::run(const RobotState& state,
         result_world.angular());
 
     // Use default constraints. Planning should be in charge of enforcing
-    // constraints on the path, here we just follow it.
+    // constraints on the trajectory, here we just follow it.
     // TODO(Kyle): Use this robot's constraints here.
```
Comment (oswinso) [inline soccer/motion/MotionControl.cpp]: I know this isn't part of this PR but issue number plz

File: soccer/planning/Instant.hpp [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+#include <time.hpp>
+
+#include "DebugDrawer.hpp"
+#include "Utils.hpp"
+#include "planning/DynamicObstacle.hpp"
+
+namespace Planning {
+
+/**
+ * This class represents a robot's motion "state" at a given time, including
```
Comment (oswinso) [inline soccer/planning/Instant.hpp]: brief

File: soccer/planning/Instant.hpp [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+    friend std::ostream& operator<<(std::ostream& stream,
+                                    const LinearMotionInstant& instant) {
+        return stream << "LinearMotionInstant(position=" << instant.position
+                      << ", velocity=" << instant.velocity << ")";
+    }
+};
+
+/**
+ * Represents the current state of a robot in a planned trajectory.
```
Comment (oswinso) [inline soccer/planning/Instant.hpp]: brief

File: soccer/planning/TrajectoryUtils.hpp [hunk 5]
```diff
@@ -0,0 +1,42 @@
+#pragma once
+
+#include "Trajectory.hpp"
+
+namespace Planning {
+
+// TODO(Kyle): both of these functions should be able to be abstracted under a
+//  single interface templated on either a shape set or a collection of dynamic
+//  obstacles.
```
Comment (oswinso) [inline soccer/planning/TrajectoryUtils.hpp]: Issue number pls

File: soccer/planning/Trajectory.cpp [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+    RJ::Time actual_end = std::min(endTime, end_time());
+
+    // If the start and end times are identical, we are trying to grab an
+    // infinitesimal trajectory. This is weird but technically not wrong.
+    if (startTime == actual_end) {
+        return Trajectory({cursor.value()});
+    }
+
+    RobotInstantSequence result_instants;
+    // TODO(Kyle): Reserve space in result_instants.
```
Comment (oswinso) [inline soccer/planning/Trajectory.cpp]: Issue num

**Review by oswinso — CHANGES_REQUESTED** [review]
Preliminary changes

File: common/Utils.hpp (hunk 1 above)
Comment (kylestach) [reply common/Utils.hpp]: O(1) instead of O(n), lmao. This should really just be an `std::remainder` though.

File: soccer/motion/MotionControl.cpp (hunk 2 above)
Comment (kylestach) [reply soccer/motion/MotionControl.cpp]: #1500

File: soccer/planning/Trajectory.hpp [hunk 7]
```diff
@@ -0,0 +1,420 @@
...
+     * @brief Get a reference to the ith instant.
+     *
+     * @param i the instant index to retrieve. Must be in range
+     *  [0, num_instants())
+     * @return a reference to the ith instant.
+     *
+     * @details Intended for use when editing a trajectory in-place; for other
+     *  cases it may be simpler to use the iterator or Cursor interfaces.
+     */
+    [[nodiscard]] auto& instant_at(int i) { return instants_.at(i); }
```
Comment (oswinso) [inline soccer/planning/Trajectory.hpp]: Why the `auto` in these functions?

File: soccer/planning/TrajectoryUtils.cpp [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+    const auto& start_hits = obstacles.hitSet(cursor.value().position());
+    while (cursor.has_value()) {
+        RobotInstant instant = cursor.value();
+
+        // Calculate the list of hits.
+        // TODO(Kyle): We could save a lot of allocation and smart pointer
+        //  reference counting overhead by making hitSet return a custom
+        //  iterator type. Then, we could use that to construct a set at the
+        //  beginning but just iterate over it in standard fashion here.
```
Comment (oswinso) [inline soccer/planning/TrajectoryUtils.cpp]: issue number

File: soccer/planning/TrajectoryUtils.hpp (hunk 5 above)
Comment (kylestach) [reply soccer/planning/TrajectoryUtils.hpp]: #1501

**Review by oswinso — CHANGES_REQUESTED** [review]
more things

File: common/Utils.hpp (hunk 1 above)
Comment (oswinso) [reply common/Utils.hpp]: I'm pretty sure with the old one the `while` acts more like an `if` though... but I think it's fine. Just a bit worried about the performance of calling `sin`, `cos` and `atan2`

Comment (kylestach) [reply common/Utils.hpp]: Remainder is preferable to any of the above. I was meaning to swap that out but this was the first change I thought of (I just really don't like the idea that we could take an unbounded amount of time in this stupid little function, especially if something hits infinity or a really really large number).

Comment (kylestach) [reply common/Utils.hpp]: Alright `std::remainder` has the correct behavior, I'll swap out to that.

Comment (kylestach) [reply common/Utils.hpp]: I didn't catch that we were doing a remainder beforehand, my bad. That should already be correct (always between -pi and pi), and we shouldn't even have needed the loops.

**Review by ghost — APPROVED** [review]

File: soccer/planning/Trajectory.cpp [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+}
+
+void Trajectory::ScaleDuration(RJ::Seconds final_duration) {
+    // Note: even though this modifies the trajectory, it does not break
+    // angle planning and so we do not need to mark the angle profile as
+    // invalid.
+    ScaleDuration(final_duration, begin_time());
+}
+
+void Trajectory::ScaleDuration(RJ::Seconds final_duration,
```
Comment (ghost) [inline soccer/planning/Trajectory.cpp:69]: I forgot where this is used, but this will be a bit slower than the old version. It's probably fine though

File: soccer/planning/Trajectory.hpp [hunk 10]
```diff
@@ -0,0 +1,422 @@
...
+class Trajectory {
+public:
+    Trajectory() = default;
+    ~Trajectory() = default;
+
+    /**
+     * @brief Create a trajectory from several "instants", each with a pose,
+     * velocity, and timestamp. This
+     */
+    explicit Trajectory(RobotInstantSequence instants)
```
Comment (ghost) [inline soccer/planning/Trajectory.hpp:41]: should use an r-value reference or a smart pointer in the function parameter. I think this makes a copy

File: soccer/planning/Trajectory.hpp [hunk 11]
```diff
@@ -0,0 +1,422 @@
...
+        : instants_(std::move(instants)) {}
+
+    /**
+     * @brief Create a trajectory from two other trajectories.
+     *  Assumes a.last() == b.first() so b.first() is skipped.
+     *
+     * @details Is able to move out of the first trajectory, but not the second.
+     */
+    Trajectory(Trajectory a, const Trajectory& b);
```
Comment (ghost) [inline soccer/planning/Trajectory.hpp:50]: Maybe have an r-value reference / smart pointer version of this function, so we can move out of the first Trajectory. Here we are copying both.

File: soccer/planning/Trajectory.hpp (hunk 10 above)
Comment (kylestach) [reply soccer/planning/Trajectory.hpp:41]: To pass into this you can do `Trajectory(std::move(instants_arg))` and it will move `instants_arg` into the `instants` argument (call the move constructor for the parameter. This is generally preferred because you can pass either a const reference or an rvalue.

File: soccer/planning/Trajectory.cpp (hunk 9 above)
Comment (kylestach) [reply soccer/planning/Trajectory.cpp:69]: It's not actually used anywhere, I think. Yeah it'll be O(n) instead of O(1) but it's super infrequently used.

File: soccer/planning/Trajectory.hpp (hunk 11 above)
Comment (kylestach) [reply soccer/planning/Trajectory.hpp:50]: See above

File: soccer/planning/Trajectory.cpp (hunk 9 above)
Comment (ghost) [reply soccer/planning/Trajectory.cpp:69]: Yeah that's fine. Trajectories are smaller now. I didn't realize that when making this comment so this is not really an issue.

File: soccer/planning/Trajectory.cpp [hunk 12]
```diff
@@ -0,0 +1,344 @@
...
+    : Cursor{trajectory, trajectory.instants_.begin()} {}
+
+Trajectory::Cursor::Cursor(const Trajectory& trajectory, RJ::Time start_time)
+    : Cursor{trajectory} {
+    seek(start_time);
+}
+
+RobotInstant Trajectory::Cursor::value() const {
+    RobotInstant instant_0 = *iterator_;
+    if (instant_0.stamp == time_) {
```
Comment (JNeiger) [inline soccer/planning/Trajectory.cpp:259]: `*iterator_.stamp` and `time_` are only equal if it's the last instant in the vector right?

File: soccer/planning/Trajectory.hpp [hunk 13]
```diff
@@ -0,0 +1,422 @@
...
+    /**
+     * @brief Stamp this trajectory for completion at the specified time.
+     */
+    void stamp(RJ::Time time) { creation_stamp_ = time; }
+
+    /*
+     * @brief Allows seeking to arbitrary positions in a Trajectory and
+     *  incrementing by arbitrary steps.
+     */
+    class Cursor {
```
Comment (JNeiger) [inline soccer/planning/Trajectory.hpp:301]: This header and source file are pretty large, would it be simple to split them up?

File: soccer/planning/TrajectoryUtils.cpp [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+    Trajectory::Cursor cursor = trajectory.cursor(start_time);
+
+    // If the trajectory has already ended, we don't need to check it.
+    if (!cursor.has_value()) {
+        return false;
+    }
+
+    // Limit iterations to 400. This will continue to operate at dt = 0.05 until
+    // we hit a 20 second trajectory. If our trajectory is longer than that,
+    // something is probably wrong.
```
Comment (JNeiger) [inline soccer/planning/TrajectoryUtils.cpp]: You should specify that the dt increases after 20 seconds as well

File: common/Geometry2d/Pose.hpp [hunk 15]
```diff
@@ -128,12 +128,38 @@ class Pose {
...
+     */
+    bool operator==(const Pose& other) const {
+        return _position == other._position && _heading == other._heading;
+    }
+
+    /**
+     * Inequality comparison operation.
+     */
+    bool operator!=(const Pose& other) const {
+        return _position != other._position || _heading != other._heading;
```
Comment (JNeiger) [inline common/Geometry2d/Pose.hpp]: Since you have the equality operator, you can just do `!(*this == other)`. This reduces copy paste errors

**Review by JNeiger — COMMENTED** [review]
Just a few things here and there

File: soccer/planning/Trajectory.hpp (hunk 13 above)
Comment (kylestach) [reply soccer/planning/Trajectory.hpp:301]: It would be a bit annoying, `Cursor` and `Trajectory::Cursor` are pretty much interdependent.

File: soccer/planning/Trajectory.cpp (hunk 12 above)
Comment (kylestach) [reply soccer/planning/Trajectory.cpp:259]: Yes. This check is necessary to avoid dereferencing the `end()` iterator in the case that we're pointing to the end.

File: common/Geometry2d/Pose.hpp [hunk 16]
```diff
@@ -289,12 +313,38 @@ class Twist {
...
+     * Equality comparison operation.
+     */
+    bool operator==(const Twist& other) const {
+        return _linear == other._linear && _angular == other._angular;
+    }
+
+    /**
+     * Inequality comparison operation.
+     */
+    bool operator!=(const Twist& other) const {
```
Comment (JNeiger) [inline common/Geometry2d/Pose.hpp]: `!=` oeprator that you can swap to `!( == )`

File: soccer/planning/Instant.hpp [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+     */
+    bool operator==(const RobotInstant& other) const {
+        return pose == other.pose && velocity == other.velocity &&
+               stamp == other.stamp;
+    }
+
+    /**
+     * Inequality comparison operator.
+     */
+    bool operator!=(const RobotInstant& other) const {
```
Comment (JNeiger) [inline soccer/planning/Instant.hpp:88]: `!=` operator that you can swap to `!( == )`

File: soccer/planning/Trajectory.hpp (hunk 13 above)
Comment (JNeiger) [reply soccer/planning/Trajectory.hpp:301]: Nevermind then

**Review by JNeiger — COMMENTED** [review]
Missed two `!=` operators. LGTM after that

**Review by JNeiger — APPROVED** [review]

**Review by oswinso — COMMENTED** [review]
lgtm

**Review by oswinso — APPROVED** [review]
lgtm
